### DISCUSSION TITLE
Instruct LIM to ask some questions in sequence and then perform API call based on the user answers

### DISCUSSION THREAD

**Discussion by QA-TonyHe** [FlowiseAI/Flowise#2528] [Q&A]
Here is the detailed scenario:
I want to get the basic infomation of the user: name, age, gender, job.
to improve user experience, I prefer the LLM ask questions one by one to collect these data, rather than ask theuser to provide all these information at once.
after got the user basic info, I want to cell API to store this user to DB, the method I can think up is by custom function tool,

But I have trouble using the existing node,because it involves variables passing through chain and agent.

Can help work this out? thanks ahead.

**Comment by mdlmarkham** [comment]
To clarify - you want the AI to converse with the user and determine precisely what it is that they want to know, and then you want to use a tool to store this information to a database?

**Reply by Panconjugo23** [reply]
I'm also interested in that topic. Did you guys find a solution for this?

**Comment by gauravkaliadev** [comment]
Is there anything we can use in Flowise to achieve this?

**Reply by andydukes** [reply]
Hi, I responded to a similar question on discord earlier (copy below). Let me know if you still need help and I'll write some notes up. 

- I’ve literally just got this solution working after a couple of weeks configuring and tuning so that part of my flow asks the user a series of questions, receives the answers, is able to verify that the user response is appropriate, and cycle through a list. while retaining the responses, after which it can kick off actions and execute tools based on the responses and also hand on to other agents etc. 
I will try and write up some notes and tips (and some cheats) next week and drop them to you if you are still in need of help / ideas?

**Comment by gauravkaliadev** [comment]
Hi,

Thanks for responding.

I have spent a few days myself to explore the options but i couldn't find
anything promising.

Could you share your insights please.

Thanks

**Comment by andydukes** [comment]
Hi! Sure no problem. I’ll find some time at the weekend and write up my notes and get them over to you. I’d would normally be happy just to forward the json flow to you, but it’s a clients project and they are actually on of the big AI owners so it’s just stuffed with confidential bits. I’ll have it with you by Sunday. Best AndySent from my iPhoneOn 14 Nov 2024, at 03:24, Gaurav Kalia ***@***.***> wrote:﻿
Hi,



Thanks for responding.



I have spent a few days myself to explore the options but i couldn't find

anything promising.



Could you share your insights please.



Thanks

**Comment by gauravkaliadev** [comment]
Thanks, that sounds great.

I'm really looking forward to it.

Thanks

**Comment by andydukes** [comment]
Hi Gaurav,

Sorry for the delay in getting this to you.

I've attached the flow JSON for the Q&A Loop example.  My own requirement was for a list of questions to be answered, one by one, and for the user input to be checked against a schema that described valid answers to the question - so I'm sure you will be able to simplify this for you own situation. 

NOTE: Make sure you change the Agent Memory path to match your own setup.


￼


￼

**Reply by gauravkaliadev** [reply]
Thanks @andydukes.

I can't see an attachment.

**Comment by andydukes** [comment]
Hi Guys.
Lets try this way 
[AgentFlow-QA-Loop-Example-Agents.json](https://github.com/user-attachments/files/17831878/AgentFlow-QA-Loop-Example-Agents.json)
![image](https://github.com/user-attachments/assets/40a6942c-3074-4a90-b82c-1e1521226d09)

**Reply by gauravkaliadev** [reply]
Thanks @andydukes .

I will give this a go.

**Reply by erlebach** [reply]
I am getting an error when importing your JSON file (in the Marketplace, or in AgentFlow or in ChatFlow). I am using Flowise 2.2.1 on the Mac. I attach
![2024-12-05_17-05-55](https://github.com/user-attachments/assets/8ef556b8-c5ee-44cc-afec-35b472e3e247)
 an image of the error message.

**Reply by andydukes** [reply]
Hi. Sorry only just saw your replay. I think if you create and empty canvas and then import the json into the new can as it should work :)

**Reply by ahardwick** [reply]
Hey, @andydukes 

Could you also share the JSON response that comes from the custom survey tool?

Cheers

**Reply by ahardwick** [reply]
Ok, it was as a simple as I thought.

```
{
  "questions" : [ 
    {
    "inputQuestion": "Enter Name",
    "inputName": "Name",
    "schema": "Must be a name"
    },
    {
    "inputQuestion": "Enter Age",
    "inputName": "Age",
    "schema": "Age must be between 0 and 100"
    }
  ]
}
```

**Comment by jawadashraf** [comment]
Hi @andydukes ,

Thank you so much for the help.
How can I save the response to sqlite db.

Regards

**Reply by andydukes** [reply]
Hi. Add a tool node to the last LLM and add a simple tool that grabs the previous output and then either send the result somewhere or if you meant to ask how do you save the state back to the state table,  then get the existing final LLM to  output structure JSON to the state table. Hope this helps

**Reply by jawadashraf** [reply]
Thank you so much for the quick response.
I am trying to save the data to a new table that I created in the sqlite db.
Here is the testing code

```
const sqlite3 = require('sqlite3').verbose();
const db = new sqlite3.Database('database.sqlite'); // I tried with the full path as well

const scenarioValue = $flow.state.scenarioValue;
const scenarioState = $flow.state.scenarioState;

db.serialize(() => {
    db.run('
        CREATE TABLE IF NOT EXISTS my_responses (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            user_id TEXT,
            question TEXT,
            response TEXT
        )
    ');
});
```

I get the error: sqlite3 module not found, even though it exists in node_modules table

Thanks

**Comment by YuriyT2232134fdfg** [comment]
My file downloaded fine. Thank you very much. And what exactly or what actions can be performed by this script? Thanks for the reply)
![test-1](https://github.com/user-attachments/assets/0a7ec2d2-3258-4b96-b334-3106fa0ef894)
